Flatpak

mattdangerw opened the pull request:
https://phabricator.endlessm.com/T13552

mattdangerw committed:
Add a package.json

mattdangerw committed:
Initial implementation

mattdangerw committed:
Add example usage

mattdangerw committed:
Support adding symlinks in /app

mattdangerw committed:
Add debugging logging

mattdangerw committed:
Add a readme

ptomato commented on README.md:
```diff
@@ -1,2 +1,80 @@
+best for app build systems which take prebuilt binaries and files and target a
+number of package formats at once. This is common in electron and nw.js workflows.
+
+To actual build node native modules and other applications dependencies inside
```
grammar nit: actual -> actually

ptomato commented on README.md:
```diff
@@ -1,2 +1,80 @@
+line version is planned.
+
+#### Usage
+flatpak-bundler provides a single method, `bundle`, which takes in a options
```
nit: a -> an

ptomato commented on index.js:
```diff
@@ -1,6 +1,140 @@
+
+var promise = require('bluebird');
+var exec = promise.promisify(require('child_process').exec);
+var copy = promise.promisify(require('fs-extra').copy);
```
You could use `promise.promisifyAll` here or import one of the fs-extra drop-in replacements

ptomato commented on index.js:
```diff
@@ -1,6 +1,140 @@
+function flatpakBuildInit (options) {
+    var args = ['flatpak build-init'];
+    addCommandLineOption(args, 'arch', options.arch);
+    args.push(options.buildDir);
```
I would make sure to shell-quote these args before joining them, in case they contain spaces. Kind of disappointing that `child_process.exec` takes a string instead of an args array...

ptomato replied:
So it might be better in that case to use spawn

mattdangerw replied:
exec is nicer in that it waits for the child process to finish so I don't have to deal with streams

ptomato commented on index.js:
```diff
@@ -1,6 +1,140 @@
+    .then(flatpakBuildFinish)
+    .then(flatpakBuildExport)
+    .then(flatpakBuildBundle)
+    .then(function (options) {
```
I'm sure there must be a Bluebird function that converts your promise chain back into a node callback at the end :smile:

mattdangerw replied:
Indeed, there's `asCallback` or `nodeify`. Just switched to native promises though...

mattdangerw committed:
Basic desktop and icon renaming

mattdangerw committed:
Require node > 4.0.0

mattdangerw committed:
Use const and let

mattdangerw committed:
Native promises

mattdangerw committed:
Move to standard style

mattdangerw committed:
consistent promise chain style

mattdangerw committed:
Don't make everything return the options object

mattdangerw committed:
Document new API

mattdangerw committed:
Rewrite to use flatpak-builder

mattdangerw committed:
Remove example

We'll just keep it in the readme

mattdangerw committed:
Rewrite with flatpak-builder --finish-only

This doesn't actually exists yet...

mattdangerw committed:
Use spawn instead of exec

mattdangerw committed:
Allow missing runtime with new flatpak builder option

mattdangerw committed:
Use template literals

mattdangerw committed:
MIT license

mattdangerw committed:
Add cleanTmpdirs option

mattdangerw committed:
README: document coming api

mattdangerw committed:
Don't apply default to manifest options

mattdangerw committed:
Separate build-export call

We need this to be able to support custom export in the flatpak

mattdangerw committed:
Normalized id/app-id in manifest

Both are valid, but easier to only handle one in our code

mattdangerw committed:
Support custom export files

Outside the usual desktop/icons/appstream...

mattdangerw committed:
Auto install/update runtime sdk and base app

mattdangerw committed:
More robust arg handling

If we use --opt val instead of --opt=val we can better take
advantage of spawn function's auto quoting

mattdangerw committed:
Add some flatpakrefs for example code

mattdangerw committed:
package.json: version 1.0.0

semver says if we are starting a shared project start at 1.0.0, so
lets do that

mattdangerw committed:
Add .npmignore

mattdangerw committed:
Fixup getOptionsWithDefaults

Fix a number of small bugs and style nits

mattdangerw committed:
Use proper flatpak build-bundle args

One thing we were passing in was not valid

mattdangerw committed:
README: specify flatpak version requirement

mattdangerw committed:
README: link the online docs

mattdangerw committed:
README: reorder build options based on relevance

mattdangerw committed:
Convert node arch to flatpak style arch

ptomato commented on README.md:
```diff
@@ -1,2 +1,207 @@
+flatpakBundler.bundle({ // Manifest
+  id: 'org.world.Hello',
+  base: 'io.atom.electron.BaseApp', // Electron base application
+  baseFlatpakref: FIXME, // So we can auto install the runtime
```
Should this be some value to indicate auto-installing it?

mattdangerw replied:
Not sure. There's the `autoInstallBase` build option that controls this, which I default to true if `baseFlatpakref` is set in your manifest cause I figure that's what a user would generally want.

But the goal with this split is you could use the same app manifest (where you want to pull your runtimes from is unlikely to change) with different types of builds (say `autoInstallBase: false` to just use your local hacked up version).

But maybe where you want to pull your platform/sdk/base from is really more of a build option? And I should just remove these from the manifest entirely?

Another thing I was unsure of was whether to decouple the auto install and auto update options. I wanted to avoid an explosion of properties, but maybe there's some reason you'd want one but not the other.

The only thing I think is worth protecting here is avoiding any setup. A user should be able to go to any machine as long as node and flatpak are available, build their electron app as a flatpak without any further manual installation of dependencies.

ptomato commented on index.js:
```diff
@@ -0,0 +1,298 @@
+const symlink = promisify(fs.symlink)
+const tmpdir = promisify(require('tmp').dir)
+
+function kebabify (object) {
```
:joy: I did not know this term.

mattdangerw replied:
I'm pretty sure by naming functions like `kebabify` I can ingratiate myself with the node community.

ptomato commented on index.js:
```diff
@@ -0,0 +1,298 @@
+}
+
+function flatpakifyArch (arch) {
+  if (arch === 'ai32') return 'i386'
```
I think this should be `ia32`?

ptomato commented on index.js:
```diff
@@ -0,0 +1,298 @@
+
+      logger(`Using manifest...\n${JSON.stringify(manifest, null, '  ')}`)
+      logger(`Using options...\n${JSON.stringify(options, null, '  ')}`)
+    })
```
Would be a nice touch to have a `ensureFlatpakVersion` that would let you know if you are using pre-0.6.13.

mattdangerw replied:
I'm not sure how to do this robustly. Call `flatpak --version` and look for the first `#.#.#` match? Don't want to mess with pkgconfig parsing or nothing like that.
